softdevice_handler, boards: improvements to LF clock and oscillator configuration

anhmolt opened the pull request:
Improve clock and oscillator configuration. See individual commit messages for more info.

anhmolt committed:
boards: bm_nrf54l15dk: add oscillator config dtsi file with comments

Add dtsi file, with comments, for configuring HF and LF external
oscillators.
The file is to be used for configuring external/internal
load-capacitance for HF and LF oscillators. If an external oscillator
for the LF clock is not desired, the LFXO can be disabled to use
the internal LFRC.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
softdevice_handler: improve LF clock Kconfig options

* Add a Kconfig choice for LF clock source selection.
* Only allow selecting LFXO if enabled in oscillator dts.
* Only allow selecting LFXO or LFRC if GRTC is configured to the same.
* Add a Kconfig choice for the LFRC calibration accuracy setting.
* Improve the LFRC calibration Kconfigs by adding conditional
  prompts and default values dependent on the LF clock source selection.
* Add some build asserts for checking that the GRTC is configured
  appropriately for SoftDevice use.
* Correct the prompt of CONFIG_NRF_SDH_CLOCK_LF_RC_TEMP_CTIV.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

NordicBuilder commented:
The following west manifest projects have changed revision in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| nrf | https://github.com/nrfconnect/sdk-nrf/commit/a7e36dcbea5d835baea5e96ea80d0b7955e296fd | https://github.com/nrfconnect/sdk-nrf/pull/25495 | [nrfconnect/sdk-nrf#25495/files](https://github.com/nrfconnect/sdk-nrf/pull/25495/files) |

⛔ **DNM label due to: 1 project with PR revision**

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

bjorn-tore-taraldsen commented on boards/nordic/bm_nrf54l15dk/bm_nrf54l15dk_nrf54l051015_oscillator.dtsi:
```diff
@@ -0,0 +1,55 @@
+
+&grtc {
+	owned-channels = <0 1 2 3 4 5 6 7 8 9 10 11>;
+	/* Channels 7-11 reserved for Zero Latency IRQs, 3-4 for FLPR */
```
Should we modify this one to match Bare Metal/ SoftDevice use case?
- SoftDevice (MPSL) will use channel 7-11 (i.e. set as child-owned-channel)
- Update comment to reflect this (replace Zero Latency IRQ)
- We don't support FLPR in BM, i.e. should we remove channel 3 and 4 from "child-owned-channel"?

anhmolt replied:
I have added a commit to modify this. Please have a look.

bjorn-tore-taraldsen commented on subsys/softdevice_handler/nrf_sdh.c:
```diff
@@ -14,11 +14,25 @@
+BUILD_ASSERT(CONFIG_NRF_SDH_CLOCK_LF_RC_TEMP_CTIV == 0, "rc_temp_ctiv must be 0 when usings LFXO");
+BUILD_ASSERT(CONFIG_NRF_SDH_CLOCK_LF_ACCURACY == 0, "accuracy must be 0 when using LFXO");
+
+#if defined(CONFIG_NRF_GRTC_START_SYSCOUNTER) && !defined(CONFIG_NRF_GRTC_TIMER_SOURCE_LFXO)
```
Regarding SYSCOUNTER, should we add a separate WARNING ?
SoftDevice (MPSL) have the following requirements: Before enabling SoftDevice: "GRTC must be started by the application ......... the SYSCOUNTER must be enabled ......."
ref: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrfxlib/mpsl/doc/mpsl.html#:~:text=For%20the%20nRF54L%20Series%3A

anhmolt replied:
Added a build assert. Please have a look.

bjorn-tore-taraldsen reviewed (commented):
I'm missing the GRTC configuration, how/ where should we place that:
We should guide customer to select between : 
   CONFIG_NRF_GRTC_TIMER_SOURCE_LFXO=y
   CONFIG_NRF_GRTC_TIMER_SOURCE_LFLPRC=y

anhmolt committed:
boards: bm_nrf54l15dk: modify grtc to the Bare Metal use case

GRTC channels 7 to 11 are used by SoftDevice. FLPR is not
currently in use in the nRF Bare Metal and the channels
for FLPR have been allocated to application CPU.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
boards: bm_nrf54l15dk: explicitly set the GRTC timer source to LFXO

Explicitly set the GRTC timer source to LFXO in the board defconfig.
Also add the associated LFLPRC choice, but commented out.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt commented on subsys/softdevice_handler/nrf_sdh.c:
```diff
@@ -14,11 +14,25 @@
+
+#if defined(CONFIG_NRF_GRTC_START_SYSCOUNTER) && defined(CONFIG_NRF_GRTC_TIMER_SOURCE_LFXO)
+#warning Please select CONFIG_NRF_GRTC_TIMER_SOURCE_LFLPRC or \
+	 CONFIG_NRF_GRTC_TIMER_SOURCE_SYSTEM_LFCLK when using CONFIG_NRF_SDH_CLOCK_LF_SRC_RC
```
Okay. I have removed this.

eivindj-nordic commented on west.yml:
```diff
@@ -14,7 +14,7 @@ manifest:
     - name: nrf
       repo-path: sdk-nrf
-      revision: a7e36dcbea5d835baea5e96ea80d0b7955e296fd
+      revision: 8316fd95485760c756caa976947801623247447d
```
sdk-nrf revision updated in separate PR.

eivindj-nordic commented:
Would be good with a changelog entry and some documentation for this :)

anhmolt commented on doc/nrf-bm/release_notes/release_notes_changelog.rst:
```diff
@@ -42,9 +42,15 @@ SoftDevice Handler
+    The default GRTC timer source of the ``bm_nrf54l15dk`` board is LFXO :kconfig:option:`CONFIG_NRF_GRTC_TIMER_SOURCE_LFXO`.
+    Select the :kconfig:option:`CONFIG_NRF_GRTC_TIMER_SOURCE_LFLPRC` Kconfig option if it is desired to run the GRTC from RC.
+  * The :kconfig:option:`CONFIG_NRF_SDH_CLOCK_LF_RC_CTIV` and :kconfig:option:`CONFIG_NRF_SDH_CLOCK_LF_RC_TEMP_CTIV` Kconfig options with ranges, default values, help text, and conditional prompt.
+    These Kconfig options are zero if the LF clock source is XO and are only user configurable when the LF clock source is RC.
```
Fixed. Thanks.

peknis reviewed (approved):
Approved with a nit.
